### ISSUE TITLE
Error: The driver doesn't support the version of ODBC behavior that the application requested (see SQLSetEnvAttr)

### ISSUE THREAD

**Issue by fdcastel** [pacman82/odbc2parquet#637]
I'm getting this error when trying to use the latest Firebird ODBC driver.

**What ODBC specification version does odbc2parquet require?**

Related upstream issue: https://github.com/FirebirdSQL/firebird-odbc-driver/issues/227

**Comment by pacman82** [comment]
It announces to ODBC that it uses 3.81. However if your driver and driver manger can only handle 3.5, you'll be fine too, apart from a warning.

The upstream issue claims odbc 4 is released. This would be news to me, though.

**Comment by fdcastel** [comment]
Thank you, @pacman82. 

If there is no harm, I believe the best practice would be to announce the lowest version needed by the code (exactly to avoid problems like this).

According to [an old blog post from Microsoft](https://learn.microsoft.com/en-us/archive/blogs/dataaccesstechnologies/a-new-release-of-odbc-for-modern-data-stores), ODBC 4 was released around the time of Microsoft SQL Server 2017. However, it is hard to find a reliable source for this, since the drivers documentation often do not inform the exact version of the specification they adhere to.

**Comment by fdcastel** [comment]
According to [`SQLSetEnvAttr` documentation](https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlsetenvattr-function) the only difference between `SQL_OV_ODBC3_80` and `SQL_OV_ODBC3`  is that the driver supports [C Data Types in ODBC](https://learn.microsoft.com/en-us/sql/odbc/reference/develop-app/c-data-types-in-odbc?view=sql-server-ver16).

Do you need this feature? If not, it's probably safe to require `SQL_OV_ODBC3`.

P.S.: It's hard to find detailed information about each specification version, but I'm assuming that `SQL_OV_ODBC3` corresponds to what Microsoft docs call the '3.5' specification.

**Comment by pacman82** [comment]
For the current feature set 3.5 is likely enough. If I ever want to maybe cancel running queries (e.g. if the user closes the tool with ctrl-c), or some fancy features like that, 3.8 would be helpful though.

Please help me understand the severity of the situation better. I've seen a diagnostic record with some older drivers only supporting 3.5, but so far at least all it yielded has been a warning. Does it actually error out for you, with your driver?

**Comment by fdcastel** [comment]
> Please help me understand the severity of the situation better. I've seen a diagnostic record with some older drivers only supporting 3.5, but so far at least all it yielded has been a warning. Does it actually error out for you, with your driver?

Yes. It gives me the error message and aborts. I cannot export any data.

**Comment by fdcastel** [comment]
Please let me know if there are any additional tests or logs I can provide to help you gain a better understanding of the problem.

**Comment by pacman82** [comment]
It would be helpful if you could post the entire output to the command line, while running odbc2parquet with the -vvv flag to trigger very verbose output.

Are you running on Windows? If so, what version?

**Comment by pacman82** [comment]
`odbc2parquet 6.1.0` has been released, which utilizes ODBC 3.5

**Comment by fdcastel** [comment]
Thanks @pacman82! Much appreciated.

> It would be helpful if you could post the entire output to the command line, while running odbc2parquet with the -vvv flag to trigger very verbose output.

Sure! 

Using odbc2parquet v6.0.7:
```powershell
.\odbc2parquet.exe -vvv query --connection-string "Driver={Firebird ODBC Driver};Database=/db/test.fdb;UID=SYSDBA;PWD=masterkey;" /temp/relations.parquet  "SELECT * FROM rdb`$relations;"

2024-09-19T13:48:39-03:00 - DEBUG ODBC Environment created.
2024-09-19T13:48:39-03:00 - DEBUG SQLAllocHandle allocated connection (Dbc) handle '0x1a4f41563e0'
2024-09-19T13:48:40-03:00 - WARN State: 01000, Native error: 0, Message: [Microsoft][ODBC Driver Manager] The driver doesn't support the version of ODBC behavior that the application requested (see SQLSetEnvAttr).
2024-09-19T13:48:40-03:00 - INFO Database Management System Name: Firebird
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 1: ColumnDescription { name: [82, 68, 66, 36, 86, 73, 69, 87, 95, 66, 76, 82], data_type: LongVarbinary { length: Some(2147483647) }, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 1: Binary { length: 2147483647 }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 2: ColumnDescription { name: [82, 68, 66, 36, 86, 73, 69, 87, 95, 83, 79, 85, 82, 67, 69], data_type: LongVarchar { length: Some(2147483647) }, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 2: WText { max_str_len: 2147483647 }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 3: ColumnDescription { name: [82, 68, 66, 36, 68, 69, 83, 67, 82, 73, 80, 84, 73, 79, 78], data_type: LongVarchar { length: Some(2147483647) }, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 3: WText { max_str_len: 2147483647 }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 4: ColumnDescription { name: [82, 68, 66, 36, 82, 69, 76, 65, 84, 73, 79, 78, 95, 73, 68], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 4: I32 { nullable: true }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 5: ColumnDescription { name: [82, 68, 66, 36, 83, 89, 83, 84, 69, 77, 95, 70, 76, 65, 71], data_type: SmallInt, nullability: NoNulls }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 5: I32 { nullable: false }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 6: ColumnDescription { name: [82, 68, 66, 36, 68, 66, 75, 69, 89, 95, 76, 69, 78, 71, 84, 72], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 6: I32 { nullable: true }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 7: ColumnDescription { name: [82, 68, 66, 36, 70, 79, 82, 77, 65, 84], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 7: I32 { nullable: true }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 8: ColumnDescription { name: [82, 68, 66, 36, 70, 73, 69, 76, 68, 95, 73, 68], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 8: I32 { nullable: true }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 9: ColumnDescription { name: [82, 68, 66, 36, 82, 69, 76, 65, 84, 73, 79, 78, 95, 78, 65, 77, 69], data_type: WChar { length: Some(31) }, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 9: WText { max_str_len: 62 }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 10: ColumnDescription { name: [82, 68, 66, 36, 83, 69, 67, 85, 82, 73, 84, 89, 95, 67, 76, 65, 83, 83], data_type: WChar { length: Some(31) }, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 10: WText { max_str_len: 62 }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 11: ColumnDescription { name: [82, 68, 66, 36, 69, 88, 84, 69, 82, 78, 65, 76, 95, 70, 73, 76, 69], data_type: Varchar { length: Some(255) }, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 11: WText { max_str_len: 510 }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 12: ColumnDescription { name: [82, 68, 66, 36, 82, 85, 78, 84, 73, 77, 69], data_type: LongVarbinary { length: Some(2147483647) }, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 12: Binary { length: 2147483647 }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 13: ColumnDescription { name: [82, 68, 66, 36, 69, 88, 84, 69, 82, 78, 65, 76, 95, 68, 69, 83, 67, 82, 73, 80, 84, 73, 79, 78], data_type: LongVarbinary { length: Some(2147483647) }, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 13: Binary { length: 2147483647 }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 14: ColumnDescription { name: [82, 68, 66, 36, 79, 87, 78, 69, 82, 95, 78, 65, 77, 69], data_type: WChar { length: Some(31) }, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 14: WText { max_str_len: 62 }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 15: ColumnDescription { name: [82, 68, 66, 36, 68, 69, 70, 65, 85, 76, 84, 95, 67, 76, 65, 83, 83], data_type: WChar { length: Some(31) }, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 15: WText { max_str_len: 62 }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 16: ColumnDescription { name: [82, 68, 66, 36, 70, 76, 65, 71, 83], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 16: I32 { nullable: true }
2024-09-19T13:48:40-03:00 - DEBUG ODBC column description for column 17: ColumnDescription { name: [82, 68, 66, 36, 82, 69, 76, 65, 84, 73, 79, 78, 95, 84, 89, 80, 69], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:48:40-03:00 - DEBUG ODBC buffer description for column 17: I32 { nullable: true }
2024-09-19T13:48:40-03:00 - INFO Memory usage per row is 15032387274 bytes. This excludes memory directly allocated by the ODBC driver.
2024-09-19T13:48:40-03:00 - DEBUG SQLFreeHandle dropped 0x1a4f4157c30 of type Stmt.
2024-09-19T13:48:40-03:00 - DEBUG SQLFreeHandle dropped 0x1a4f41563e0 of type Dbc.
2024-09-19T13:48:40-03:00 - DEBUG SQLFreeHandle dropped 0x1a4f4151b80 of type Env.
Error: Memory required to hold a single row is larger than the limit. Memory Limit: 2147483648 bytes, Memory per row: 15032387274 bytes.
You can use either '--batch-size-row' or '--batch-size-mib' to raise the limit. You may also try more verbose output to see which columns require so much memory and consider casting them into something smaller.
```

Using odbc2parquet v6.1.0
```powershell
.\odbc2parquet.exe -vvv query --connection-string "Driver={Firebird ODBC Driver};Database=/db/test.fdb;UID=SYSDBA;PWD=masterkey;" /temp/relations.parquet  "SELECT * FROM rdb`$relations;"

2024-09-19T13:49:45-03:00 - DEBUG ODBC Environment created.
2024-09-19T13:49:45-03:00 - DEBUG SQLAllocHandle allocated connection (Dbc) handle '0x1eae8eb63e0'
2024-09-19T13:49:45-03:00 - INFO Database Management System Name: Firebird
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 1: ColumnDescription { name: [82, 68, 66, 36, 86, 73, 69, 87, 95, 66, 76, 82], data_type: LongVarbinary { length: Some(2147483647) }, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 1: Binary { length: 2147483647 }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 2: ColumnDescription { name: [82, 68, 66, 36, 86, 73, 69, 87, 95, 83, 79, 85, 82, 67, 69], data_type: LongVarchar { length: Some(2147483647) }, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 2: WText { max_str_len: 2147483647 }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 3: ColumnDescription { name: [82, 68, 66, 36, 68, 69, 83, 67, 82, 73, 80, 84, 73, 79, 78], data_type: LongVarchar { length: Some(2147483647) }, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 3: WText { max_str_len: 2147483647 }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 4: ColumnDescription { name: [82, 68, 66, 36, 82, 69, 76, 65, 84, 73, 79, 78, 95, 73, 68], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 4: I32 { nullable: true }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 5: ColumnDescription { name: [82, 68, 66, 36, 83, 89, 83, 84, 69, 77, 95, 70, 76, 65, 71], data_type: SmallInt, nullability: NoNulls }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 5: I32 { nullable: false }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 6: ColumnDescription { name: [82, 68, 66, 36, 68, 66, 75, 69, 89, 95, 76, 69, 78, 71, 84, 72], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 6: I32 { nullable: true }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 7: ColumnDescription { name: [82, 68, 66, 36, 70, 79, 82, 77, 65, 84], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 7: I32 { nullable: true }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 8: ColumnDescription { name: [82, 68, 66, 36, 70, 73, 69, 76, 68, 95, 73, 68], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 8: I32 { nullable: true }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 9: ColumnDescription { name: [82, 68, 66, 36, 82, 69, 76, 65, 84, 73, 79, 78, 95, 78, 65, 77, 69], data_type: WChar { length: Some(31) }, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 9: WText { max_str_len: 62 }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 10: ColumnDescription { name: [82, 68, 66, 36, 83, 69, 67, 85, 82, 73, 84, 89, 95, 67, 76, 65, 83, 83], data_type: WChar { length: Some(31) }, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 10: WText { max_str_len: 62 }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 11: ColumnDescription { name: [82, 68, 66, 36, 69, 88, 84, 69, 82, 78, 65, 76, 95, 70, 73, 76, 69], data_type: Varchar { length: Some(255) }, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 11: WText { max_str_len: 510 }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 12: ColumnDescription { name: [82, 68, 66, 36, 82, 85, 78, 84, 73, 77, 69], data_type: LongVarbinary { length: Some(2147483647) }, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 12: Binary { length: 2147483647 }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 13: ColumnDescription { name: [82, 68, 66, 36, 69, 88, 84, 69, 82, 78, 65, 76, 95, 68, 69, 83, 67, 82, 73, 80, 84, 73, 79, 78], data_type: LongVarbinary { length: Some(2147483647) }, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 13: Binary { length: 2147483647 }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 14: ColumnDescription { name: [82, 68, 66, 36, 79, 87, 78, 69, 82, 95, 78, 65, 77, 69], data_type: WChar { length: Some(31) }, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 14: WText { max_str_len: 62 }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 15: ColumnDescription { name: [82, 68, 66, 36, 68, 69, 70, 65, 85, 76, 84, 95, 67, 76, 65, 83, 83], data_type: WChar { length: Some(31) }, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 15: WText { max_str_len: 62 }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 16: ColumnDescription { name: [82, 68, 66, 36, 70, 76, 65, 71, 83], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 16: I32 { nullable: true }
2024-09-19T13:49:45-03:00 - DEBUG ODBC column description for column 17: ColumnDescription { name: [82, 68, 66, 36, 82, 69, 76, 65, 84, 73, 79, 78, 95, 84, 89, 80, 69], data_type: SmallInt, nullability: Nullable }
2024-09-19T13:49:45-03:00 - DEBUG ODBC buffer description for column 17: I32 { nullable: true }
2024-09-19T13:49:45-03:00 - INFO Memory usage per row is 15032387274 bytes. This excludes memory directly allocated by the ODBC driver.
2024-09-19T13:49:45-03:00 - DEBUG SQLFreeHandle dropped 0x1eae8eb78d0 of type Stmt.
2024-09-19T13:49:45-03:00 - DEBUG SQLFreeHandle dropped 0x1eae8eb63e0 of type Dbc.
2024-09-19T13:49:45-03:00 - DEBUG SQLFreeHandle dropped 0x1eae8eb1b80 of type Env.
Error: Memory required to hold a single row is larger than the limit. Memory Limit: 2147483648 bytes, Memory per row: 15032387274 bytes.
You can use either '--batch-size-row' or '--batch-size-mib' to raise the limit. You may also try more verbose output to see which columns require so much memory and consider casting them into something smaller.
```

So, **using odbc2parquet v6.1.0 fixed the problem** (warning, actually) with the ODBC specification version. 

However, now I see that this was actually NOT the root problem. 😞 


Originally, I neglected the message 'Error: Memory required to hold a single row (...)' assuming it was caused by the previous error (ODBC version incompatible).

But now I can see that THIS was the actual problem stopping the program. As you correctly stated in a previous message, the ODBC version was just a warning, not a fatal error. 

I will open another issue for this.



> Are you running on Windows? If so, what version?

Yes. Windows 11 Build 22621 (v10.0.22621)
